### PR TITLE
[train] Simplify `ray.train.xgboost/lightgbm` (2/n): Re-implement `XGBoostTrainer` as a lightweight `DataParallelTrainer`

### PR DESCRIPTION
**Pull request by justinvyu**
## TL;DR

This PR re-implements `XGBoostTrainer` as a `DataParallelTrainer` that does not use `xgboost_ray` under the hood, in an effort to unify the trainer implementations and remove that external dependency.

## Motivation

1. **High maintenance burden** that requires a release process every time incompatibilities come up between Ray and the latest `xgboost_ray`/`lightgbm_ray` version.
    * For example, see the last few months of commit history.
    * Here’s what happens: a change in Ray (e.g. a Ray Tune deprecation/moved package) causes the latest release of `xgboost_ray` to no longer work →  `ray.train.xgboost.XGBoostTrainer` breaks → Ray Train team needs to patch a fix in this separate package, make a release, then update the pinned package version in CI.
    * This process is all manual, so this entire process including getting CI to pass takes at least 2 hours.
2. **Reducing code complexity:** Directly using `xgboost_ray` introduces significant code complexity.
`XGBoostTrainer` and `LightGBMTrainer` are data parallel trainers, but go through a completely different code path as `DataParallelTrainer` implementations.
    * After this migration, all Ray Train entrypoints will be using the `DataParallelTrainer` execution logic.
3. **Usability:** The current xgboost and lightgbm trainers are hard to use due to being a pass-through API shell on top of the xgboost.train API that people are familiar with.
    * Let’s use this opportunity to cut down these bulky external packages to a simple, lightweight integration in Ray Train, where users need to change minimal code to distribute their workload.
    * This is the same motivation as the `TorchTrainer` unification effort.
4. Minor point: Removing duplicate logic
    * `xgboost_ray` and `lightgbm_ray` are designed to be run independently, so they implement their execution loop with resource scheduling logic and error handling. There is a huge overlap in the external libraries and Tune, and it’s very difficult to navigate between the 2 codebases as a maintainer.
5. [See more reasons here.](https://github.com/ray-project/ray/pull/42767#issuecomment-1915765354)

## PR Summary

1. Introduce simplified `ray.train.xgboost.v2.XGBoostTrainer` and `ray.train.lightgbm.v2.LightGBMTrainer` that do not depend on `xgboost_ray` and `lightgbm_ray`.
    * These will have a different API compared to the existing trainers due to being subclasses of `DataParallelTrainer`. Users are able to pass in their own training function.
2. Re-implement the existing `ray.train.xgboost.XGBoostTrainer` and `ray.train.lightgbm.LightGBMTrainer` on top of the v2 counterparts.
    * **We are not going to force an API migration immediately -- let's wait for this new implementation to stabilize for 1 release before making that decision.**
3. Remove `xgboost_ray` and `lightgbm_ray` dependencies from Ray Train starting immediately from 2.10. (Will do in a follow-up PR.)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by woshiyyya** [comment]
Very neat solution. Some questions:
- Would we support or remove support for elastic training? (not sure how many users are using it, but my intuition is we are removing this feature)
- Shall we provide a simple API for users to get a `xgboost.DMatrix`. I noticed that in xgboost_ray, there will be no further processing logics in the `train_func` after passing the `RayDMatrix` into the `train()`.
- What would be the checkpointing code looks like? Will that be defining a Xgboost callback, then call `ray.train.report` in it?

**Comment by justinvyu** [comment]
@woshiyyya

> Would we support or remove support for elastic training? (not sure how many users are using it, but my intuition is we are removing this feature)
* This would remove the elastic training part. I'm not actually sure if the old `XGBoostTrainer` even allows usage of the elastic training feature. `xgboost_ray` re-implements the execution loop, which waits for distributed worker futures and launching workers when one fails.
* However, this would get caught by Ray Train, the entire placement group would get removed and restarted. OR, this would hang if the user includes a `ray.train.report` in our provided callback.
* We can revisit xgb elastic training in the future along with other data parallel trainers.

> Shall we provide a simple API for users to get a xgboost.DMatrix. I noticed that in xgboost_ray, there will be no further processing logics in the train_func after passing the RayDMatrix into the train().

* I think it's not needed to add another API on top of DMatrix. `RayDMatrix` is an abstraction used to shard the dataset across multiple Ray workers, but we don't really need that since Ray Data can do that already, so I don't want to keep RayDMatrix around.
* [Here's what regular `xgboost.DMatrix` usage looks like.](https://github.com/ray-project/xgboost_ray/blob/7cf815bb1a948f8fccc60cc0ec91da85452d3a68/xgboost_ray/examples/simple.py#L12-L17)
* Here's what it looks like with Ray Train + Ray Data. It matches up with the native way of using xgboost.
```python
           train_ds = ray.train.get_dataset_shard("train")
            train_df = train_ds.to_pandas()
            X, y = train_df.drop("y", axis=1), train_df["y"]
            dtrain = xgboost.DMatrix(X, label=y)
```

> What would be the checkpointing code looks like? Will that be a Xgboost's native callback, which calls ray.train.report on some hooks?

Two options:
1. `TuneReportCheckpointCallback` which is already our recommendation, and users can implement their own if needed similar to lightning/transformers.
2. Call `ray.train.report` manually by iteratively training more models:
```python
bst_model = None
num_boost_rounds_per_iter =
for i in range(num_iters):
    bst_model = xgboost.train(
        ..., xgb_model=bst_model,  # start from bst_model
        num_boost_rounds=num_boost_rounds_per_iter
    )
    ray.train.report(..., checkpoint=...)
```

**Comment by justinvyu** [comment]
Here's a summary of the enhancements achieved by this proposal:

| Feature          | Status | Notes |
|------------------|--------|-------|
| Elastic training | ↔️   |  Elastic training implementation is no longer attached to `XGBoostTrainer`, but this is not technically a regression, since it didn't work before anyways.   |
| Checkpointing    | ↗️     |  Still the same as before, except checkpoint loading / post-processing logic is even more flexible now. Before, everything had to be through the xgboost callback.     |
| Ray Data Integration    | ↗️     |  Previously, the integration was mostly done on the `RayDMatrix` level instead of the existing `DataParallelTrainer` logic. Now, the integration is unified across more trainers.     |
| Future xgboost features  | ↗️     |   There are many features that are easily accessible since the user has control of the training loop, including a [federated distributed learning backend](https://github.com/dmlc/xgboost/blob/65d7bf2dfeb4ab737e4097ab39f887150be04a92/python-package/xgboost/collective.py#L55-L62), [iterator-based DMatrix loading](https://xgboost.readthedocs.io/en/latest/tutorials/external_memory.html#data-iterator), and whatever else xgboost adds in the future. |
| Usability  | ↗️     |  The current `XGBoostTrainer` is 2 unnecessary layers on top of the native `xgboost.train` API that people are familiar with. You have to pass configs through `XGBoostTrainer(params, label_column, **train_kwargs)`, which are then passed to `xgboost_ray.train`, which also has a bunch of arguments passed through with `**kwargs`. This is really had to use and can't utilize editor auto-complete. It's easier to let the user call `xgboost.train` directly. |

Let me know if there are any "regressions" that I'm missing with this change.

File: python/ray/train/xgboost/config.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    def backend_cls(self):
+        return _XGBoostBackend
+
+
+class _XGBoostBackend(Backend):
+    def on_start(self, worker_group: WorkerGroup, backend_config: XGBoostConfig):
+        user_args = backend_config.args or {}
+        xgboost_communicator = user_args.pop(
+            "xgboost_communicator", backend_config.xgboost_communicator
+        )
```
Comment (matthewdeng) [inline python/ray/train/xgboost/config.py:35]: Why is this in the `user_args`?

File: python/ray/train/xgboost/config.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        if xgboost_communicator == "rabit":
+            # Set up the rabit tracker on the Train driver.
+            num_workers = len(worker_group)
+            self.rabit_args = {"DMLC_NUM_WORKER": num_workers}
+            train_driver_ip = ray.util.get_node_ip_address()
```
Comment (matthewdeng) [inline python/ray/train/xgboost/config.py:40]: Should this be this IP or the rank 0 worker?

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 3]
```diff
@@ -122,3 +124,94 @@ def preprocess_datasets(self) -> None:
...
+
+        from xgboost.collective import CommunicatorContext
+
+        with CommunicatorContext():
+            yield
+
+    def train_fn(config):
+        from xgboost.collective import CommunicatorContext
+
+        with CommunicatorContext():
```
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:184]: Is this needed?

**Review by matthewdeng — COMMENTED** [review]
So clean!

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:184]: Yup, this is the thing that actually connects the worker to the collective group (kind of like init_process_group).

I was trying to do it for the user [here](https://github.com/ray-project/ray/pull/42767/files/d8ecfa8ffefa0c1da25c4b6540f993e12a18d1c7#diff-e40514dc5fac1d96905235c5090fbeeb747889a95a3dd86228f8737348911236R72-R78), but that didn't end up working, since the context needs to be directly wrapping the user code it seems.

Comment (matthewdeng) [reply python/ray/train/xgboost/xgboost_trainer.py:184]: Oh I see. @woshiyyya brought up a similar thing for Torch where we might want to set the torch device, which also needs to modify the user code since it has to be run in the same thread.

Maybe we can have some sort of decorator abstraction that surrounds the user's train function?

File: python/ray/train/xgboost/config.py (hunk 2 above)
Comment (justinvyu) [reply python/ray/train/xgboost/config.py:40]: The rabit process should be on the "driver", which in this case is the Trainer. All ranks connect to the driver rabit process.

Take a look at how this dask distributed xgboost test sets it up: https://github.com/dmlc/xgboost/blob/662854c7d75ef1ec543ee0db73098227de5be59c/tests/test_distributed/test_with_dask/test_with_dask.py#L1619-L1654

File: python/ray/train/xgboost/config.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/xgboost/config.py:35]: Yeah, I was debating whether to just keep this as part of the args.

The main thing is that we'd need to use this value to determine how to do setup on the workers below. For now, we can just maintain the status quo and only support `rabit`, but we can support the `federated` option in the future.

Comment (justinvyu) [reply python/ray/train/xgboost/config.py:35]: Another note:

The main thing is that we'd need to use this value to determine how to do setup on the workers below. For now, we can just maintain the status quo and only support `rabit`, but we can support the `federated` option in the future.

Comment (matthewdeng) [reply python/ray/train/xgboost/config.py:35]: Oh gotcha. I don't have a super strong opinion from my current understanding, so maybe we can just try out a few examples (e.g. what it looks like as a user who is trying to configure their setup) to see what's the most intuitive.

One suggestion I have is to move the validation logic into a `__post_init__` method of `XGBoostConfig` so it's centralized to the initialization.

File: python/ray/train/xgboost/config.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+class _XGBoostBackend(Backend):
+    def on_start(self, worker_group: WorkerGroup, backend_config: XGBoostConfig):
+        user_args = backend_config.args or {}
+        xgboost_communicator = user_args.pop(
+            "xgboost_communicator", backend_config.xgboost_communicator
+        )
+        if xgboost_communicator == "rabit":
+            # Set up the rabit tracker on the Train driver.
+            num_workers = len(worker_group)
+            self.rabit_args = {"DMLC_NUM_WORKER": num_workers}
```
Comment (matthewdeng) [inline python/ray/train/xgboost/config.py:39]: nit: Move all instance variables into an `__init__` method so they're easier to keep track of?

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 3 above)
Comment (woshiyyya) [reply python/ray/train/xgboost/xgboost_trainer.py:184]: Yes. I was trying to set torch cuda device by default, but it only works when we call it inside the training function. I am thinking we can have a function decorator, so that we can inject some environment setup in an elegant way.

e.g.
```
@ray.train(framework="torch")
def train_func():
    ...

```

File: python/ray/train/xgboost/config.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            train_driver_ip = ray.util.get_node_ip_address()
+            self.tracker = RabitTracker(host_ip=train_driver_ip, n_workers=num_workers)
+            self.rabit_args.update(self.tracker.worker_envs())
+
+            self.tracker.start(num_workers)
+
+            print("DMLC_TRACKER_ENV_START\n")
```
Comment (woshiyyya) [inline python/ray/train/xgboost/config.py:46]: Shall we replace the `print` with `logging` library?

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 3 above)
Comment (woshiyyya) [reply python/ray/train/xgboost/xgboost_trainer.py:184]: Seems that it's more crucial for the new XGBoostTrainer API. We could consider having this decorator so users don't have to think about it.

File: python/ray/train/xgboost/config.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/train/xgboost/config.py:35]: Good idea!

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 3 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:184]: Interesting, we may be able to get rid of the `Trainer`s and mirror the Ray Core API more. 😆

File: python/ray/train/_internal/data_config.py [hunk 6]
```diff
@@ -20,6 +20,7 @@ def __init__(
         self,
         datasets_to_split: Union[Literal["all"], List[str]] = "all",
         execution_options: Optional[ExecutionOptions] = None,
+        convert_to_data_iterator: bool = True,
```
Comment (woshiyyya) [inline python/ray/train/_internal/data_config.py:23]: Nit: This seems unclear for the people don't know about what data_iterator refers to. Consider rename it to `streaming_execution=True` or `materialize=False`?

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 7]
```diff
@@ -1,15 +1,220 @@
...
+            f"additional {remaining_iters} iterations (trees) in order "
+            "to achieve the target number of iterations "
+            f"({num_boost_round=})."
+        )
+
+    train_dataset = ray.train.get_dataset_shard(TRAIN_DATASET_KEY)
+    train_df = train_dataset.to_pandas()
+
+    eval_datasets = {
+        k: d for k, d in get_session().dataset_shard.items() if k != TRAIN_DATASET_KEY
```
Comment (woshiyyya) [inline python/ray/train/xgboost/xgboost_trainer.py:44]: Why use internal API here? Can we get eval datasets also with the `get_dataset_shard` API?

File: python/ray/train/_internal/data_config.py (hunk 6 above)
Comment (justinvyu) [reply python/ray/train/_internal/data_config.py:23]: Update: we're gonna have the user call [`DataIterator.materialize`](https://github.com/ray-project/ray/pull/43210) instead.

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 7 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:44]: You have to know the keys in advance to get them all with `get_dataset_shard` 😢  The user can specify any keys in `XGBoostTrainer(datasets)` so I have to access it privately this way.

Comment (woshiyyya) [reply python/ray/train/xgboost/xgboost_trainer.py:44]: Ah got it. I am thinking if we can extract the key in the init function of `XGBoostTrainer`?

Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:44]: Yeah, I can pass those keys through the partial args of this function.

Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py]: updated

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 8]
```diff
@@ -1,15 +1,224 @@
...
+    remaining_iters = num_boost_round
+    if checkpoint:
+        starting_model = RayTrainReportCallback.get_model(checkpoint)
+        starting_iter = starting_model.num_boosted_rounds()
+        remaining_iters = num_boost_round - starting_iter
+        logger.info(
+            f"Model loaded from checkpoint will train for "
+            f"additional {remaining_iters} iterations (trees) in order "
+            "to achieve the target number of iterations "
+            f"({num_boost_round=})."
```
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:39]:
```suggestion
            f"({num_boost_round})."
```

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 9]
```diff
@@ -1,15 +1,224 @@
...
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+        metadata: Optional[Dict[str, Any]] = None,
+        **train_kwargs,
+    ):
+        # TODO(justinvyu): [Deprecated] Remove in 2.11
+        if dmatrix_params != _DEPRECATED_VALUE:
+            raise DeprecationWarning(
+                "`dmatrix_params` is deprecated, since XGBoostTrainer no longer "
+                "depends on the `xgboost_ray.RayDMatrix` utility."
+            )
```
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:161]: Is there any alternative needed here for functional parity?

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 10]
```diff
@@ -1,15 +1,224 @@
...
+        scaling_config: Configuration for how to scale data parallel training.
+        run_config: Configuration for the execution of the training run.
+        resume_from_checkpoint: A checkpoint to resume training from.
+        metadata: Dict that should be made available in `checkpoint.get_metadata()`
+            for checkpoints saved from this Trainer. Must be JSON-serializable.
+        **train_kwargs: Additional kwargs passed to ``xgboost.train()`` function.
+    """
+
+    _handles_checkpoint_freq = True
+    _handles_checkpoint_at_end = True
```
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:140]: Can remove these now?

File: python/ray/train/tests/test_xgboost_trainer.py
Comment (matthewdeng) [inline python/ray/train/tests/test_xgboost_trainer.py:1]: Nice validation that this change cleans up a bunch of rough edges.

File: python/ray/train/xgboost/xgboost_trainer.py
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:1]: Can you update the PR description?

File: python/ray/train/xgboost/config.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+            for k, v in rabit_args.items():
+                os.environ[k] = str(v)
+
+            # Ranks are assigned in increasing order of the worker's task id.
+            # This task id will be sorted by increasing world rank.
+            os.environ["DMLC_TASK_ID"] = (
+                f"[xgboost.ray-rank={ray.train.get_context().get_world_rank()}]:"
+                f"{ray.get_runtime_context().get_actor_id()}"
+            )
```
[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/xgboost/config.py:87]: Nope, this format was inspired by `xgboost.dask`: https://github.com/dmlc/xgboost/blob/7cc256e246e68a6c641ecb57a138e3c8a721c55e/python-package/xgboost/dask/__init__.py#L237

The strings will just be sorted by world rank: https://github.com/ray-project/ray/pull/42767/files#diff-e40514dc5fac1d96905235c5090fbeeb747889a95a3dd86228f8737348911236R55-R60

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 8 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:39]: Oh, I just learned about this feature:

```python
var_name = 1
print(f"{var_name=}")  # -> var_name=1
```

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 12]
```diff
@@ -1,15 +1,224 @@
...
+    # NOTE: Include the training dataset in the evaluation datasets.
+    # This allows `train-*` metrics to be calculated and reported.
+    eval_dfs[TRAIN_DATASET_KEY] = train_df
+
+    train_X, train_y = train_df.drop(label_column, axis=1), train_df[label_column]
+    dtrain = xgboost.DMatrix(train_X, label=train_y)
+    evals = []
+    for eval_name, eval_df in eval_dfs.items():
+        eval_X, eval_y = eval_df.drop(label_column, axis=1), eval_df[label_column]
+        evals.append((xgboost.DMatrix(eval_X, label=eval_y), eval_name))
```
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:61]: nit: I think there's a bit of overlap/duplication in computing the train dataset here for both train and eval, but not sure if it's worth optimizing.

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 10 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:140]: Still needed since `checkpoint_at_end` / `checkpoint_frequency` still does something for `XGBoostTrainer` (it's passed into the callback). It would raise error without this since `DataParallelTrainer` doesn't allow these args to be set.

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 12 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:61]: This is what we did previously with `xgboost_ray` to get the `train-*` metrics.

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 9 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:161]: I think the closest thing would be passing these [params into the `xgboost.DMatrix(...)`.](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.DMatrix)

This would be a new feature though, since the original usage was to pass extra params as `xgboost_ray.RayDMatrix` constructor args, but none of those apply anymore.

Let's just keep it as deprecated?

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 12 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:61]: Oh, I see what you mean, should be a quick fix

File: python/ray/train/xgboost/config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (matthewdeng) [reply python/ray/train/xgboost/config.py:87]: Hmm so both here and in Dask it'll end up sorting by the rank strings rather than integers. I guess that's fine if Dask is doing it, but maybe we can update the documentation to match? (Or we can prepend some zeros to the world rank).

[the remark this answers is not in the document]
Comment (justinvyu) [reply python/ray/train/xgboost/config.py:87]: Oh yeah, string comparison will mess up if you have more than 10 workers. I do actually want the ranks to all match up, so prepending a few 0s makes sense.

File: python/ray/train/xgboost/config.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+                os.environ[k] = str(v)
+
+            # Ranks are assigned in increasing order of the worker's task id.
+            # This task id will be sorted by increasing world rank.
+            os.environ["DMLC_TASK_ID"] = (
+                f"[xgboost.ray-rank={ray.train.get_context().get_world_rank():08}]:"
+                f"{ray.get_runtime_context().get_actor_id()}"
+            )
+
+        worker_group.execute(set_xgboost_env_vars)
```
Comment (woshiyyya) [inline python/ray/train/xgboost/config.py:89]: nit: I feel our naming of backend hooks is confusing. The difference between `on_start` and `on_training_start` is not obvious. (actually it's before and after we initialized session and allocated rankings)

No changes needed here, but I think these two things could be combined together if we initialize session at the very beginning.

Comment (justinvyu) [reply python/ray/train/xgboost/config.py:89]: Yeah, I agree. These two can be combined!

File: python/ray/train/xgboost/xgboost_trainer.py [hunk 14]
```diff
@@ -1,15 +1,227 @@
...
+
+    # NOTE: Include the training dataset in the evaluation datasets.
+    # This allows `train-*` metrics to be calculated and reported.
+    evals = [(dtrain, TRAIN_DATASET_KEY)]
+
+    for eval_name, eval_df in eval_dfs.items():
+        eval_X, eval_y = eval_df.drop(label_column, axis=1), eval_df[label_column]
+        evals.append((xgboost.DMatrix(eval_X, label=eval_y), eval_name))
+
+    with CommunicatorContext():
```
Comment (matthewdeng) [inline python/ray/train/xgboost/xgboost_trainer.py:63]: Will we (eventually) move this into the `train_func_context`?

**Review by matthewdeng — APPROVED** [review]
<div><br/>via <a href="https://giphy.com/minions/">Minions</a> on <a href="https://giphy.com/gifs/minions-minions-2-rise-of-gru-StKiS6x698JAl9d6cx">GIPHY</a></div>

File: python/ray/train/xgboost/xgboost_trainer.py (hunk 14 above)
Comment (justinvyu) [reply python/ray/train/xgboost/xgboost_trainer.py:63]: Yes, I'll add that in a followup!
